src/odemis/driver/smaract.py

```diff
@@ -3084,3 +3085,1139 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+                            (func.__name__, result, SA_SIDLL.err_code.get(result, "")))
+
+        return result
+
+
+class SA_SIError(IOError):
+    def __init__(self, errno, strerror, *args, **kwargs):
+        super(SA_SIError, self).__init__(errno, strerror, *args, **kwargs)
+
+    def __str__(self):
+        return self.strerror
+
+
+class SA_SI_EventData(Union):
+    """
+    SA_SI event data is stored as this type of union (A C union used by DLL)
+    """
+    _fields_ = [
+        ("error", c_uint32),
+        ("bufferId", c_uint32),
+        ("devEventParameter", c_int32),
+        ("unused", c_int8 * 24),
+         ]
+
+
+class SA_SI_Event(Structure):
+    """
+    SA_SI Event structure (C struct used by DLL)
+    """
+    _anonymous_ = ("u",)
+    _fields_ = [
+        ("type", c_uint32),
+        ("u", SA_SI_EventData),
+        ]
+
+
+class Picoscale(model.HwComponent):
+    """
+    A driver for a SmarAct Picoscale interferometer system.
+
+    The device does not contain any actuators. Its main functionality is to provide the position for each
+    of its channels through a VA.
+
+    Attributes
+    ==========
+    .position (VA: str --> float): position in m for each channel
+    .channels (dict str --> int): maps channel name to channel number
+    .referenced (VA: str --> bool): indicates which channels have been referenced
+
+    Functions
+    =========
+    .reference: performs adjustment routine required for precise position values
+    .stop: cancels all calls to the device
+    .scan (static method): returns list of all available Picoscale controllers
+    (+ wrapper functions for Picoscale API)
+    """
+
+    def __init__(self, name, role, locator, channels, ref_on_init, precision_mode=0, *args, **kwargs):
+        """
+        name: (str)
+        role: (str)
+        locator: (str) Use "fake" for a simulator.
+            For a real device, Picoscale controllers with USB interface can be addressed with the
+            following locator syntax:
+                usb:ix:<id>
+            where <id> is the first part of a USB devices serial number which
+            is printed on the Picoscale controller.
+            If the controller has a TCP/IP connection, use:
+                network:<ip>:<port>
+            The device can also be addressed by its serial number:
+                usb:sn:<serial_number>
+                network:sn:<serial_number>
+        channels: (str --> int) dictionary mapping channel names to channel numbers
+        ref_on_init: (bool) determines if the controller should automatically reference
+            on initialization
+        precision_mode: (0 <= int <= 5) strength of digital lowpass filter, a higher level corresponds to higher
+            precision, but lower velocity. Not available on all systems.
+        """
+        model.HwComponent.__init__(self, name, role, *args, **kwargs)
+
+        if not channels:
+            raise ValueError("Needs at least 1 axis.")
```

Comment:
Please check more thoroughly `channels`. At least, it should check the channel numbers are within 0 and num_ch -1.